tools/adaparse.cpp

```diff
@@ -1,8 +1,147 @@
 #include <cstdlib>
+#include <chrono>
 #include <iostream>
 #include <string_view>
+#include <sys/stat.h>
+#include <fmt/core.h>
 #include <cxxopts.hpp>
+#include <fstream>
 #include "ada.h"
+#include "line_iterator.h"
+#ifdef _MSC_VER
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+uint64_t nano() {
+  return std::chrono::duration_cast<::std::chrono::nanoseconds>(
+             std::chrono::steady_clock::now().time_since_epoch())
+      .count();
+}
+
+
+bool print_part(std::string_view get_part, const ada::url_aggregator& url) {
+    if(get_part.size() == 4) {
+        if(get_part[0] == 'h') {
+          if(get_part == "host") {
+            fmt::print("{}\n", url.get_host());
+            return true;
+          } else if(get_part == "hash") {
+            fmt::print("{}\n", url.get_hash());
+            return true;
+          }
+        } else if(get_part[0] == 'p') {
+          if(get_part == "port") {
+            fmt::print("{}\n", url.get_port());
+            return true;
+          }
+        }
+    } else if(get_part.size() == 6) {
+        if(get_part == "origin") {
+          fmt::print("{}\n", url.get_origin());
+          return true;
+        }
+        if(get_part == "search") {
+          fmt::print("{}\n", url.get_search());
+          return true;
+        }
+    } else if(get_part.size() == 7) {
+      if(get_part[0] == 'p') {
+        if(get_part == "protocol") {
+          fmt::print("{}\n", url.get_protocol());
+          return true;
+        }
+        if(get_part == "password") {
+          fmt::print("{}\n", url.get_password());
+          return true;
+        }
+        if(get_part == "pathname") {
+          fmt::print("{}\n", url.get_pathname());
+          return true;
+        }
+
+    }
+    } else if(get_part == "hostname") {
+        fmt::print("{}\n", url.get_hostname());
+        return true;
+
+    } else if(get_part == "username") {
+        fmt::print("{}\n", url.get_username());
+        return true;
+    }
+
+    fmt::print(stderr, "\"{}\" not found\n", get_part);
+    return false;
+}
+
+int piped_file(const cxxopts::ParseResult result, std::istream& input_stream) {
```

Comment:
This won't read from `input_stream`, of course, it will read from `std::in`.

You probably want to pass a file descriptor ? `FILE*`?

The zero you pass to `read` below is explained as follows...

https://stackoverflow.com/questions/70144683/why-is-stdin-fileno-always-zero

Reply:
Thanks for all the comments, I'll get right on to it.